cloud-controller-manager/do/certificates_test.go

```diff
@@ -0,0 +1,318 @@
+			ID:   certID,
+			Type: certTypeLetsEncrypt,
+		}, newFakeOKResponse(), nil
+	}
+	lb, ok := f.store[certID]
+	if ok {
+		return lb, newFakeOKResponse(), nil
+	}
+	return nil, newFakeNotOKResponse(), newFakeNotFoundErrorResponse()
+}
+
+func (f *kvCertService) List(ctx context.Context, listOpts *godo.ListOptions) ([]godo.Certificate, *godo.Response, error) {
+	panic("not implemented")
+}
+
+func (f *kvCertService) Create(ctx context.Context, crtr *godo.CertificateRequest) (*godo.Certificate, *godo.Response, error) {
+	panic("not implemented")
+}
+
+func (f *kvCertService) Delete(ctx context.Context, certID string) (*godo.Response, error) {
+	panic("not implemented")
+}
+
+func newKVCertService(store map[string]*godo.Certificate, reflectorMode bool) kvCertService {
+	return kvCertService{
+		store:         store,
+		reflectorMode: reflectorMode,
+	}
+}
+
+func createServiceAndCert(lbID, certID, certType string) (*v1.Service, *godo.Certificate) {
+	c := &godo.Certificate{
+		ID:   certID,
+		Type: certType,
+	}
+	s := createServiceWithCert(lbID, certID)
+	return s, c
+}
+
+func createServiceWithCert(lbID, certID string) *v1.Service {
+	s := createService(lbID)
+	s.Annotations[annDOCertificateID] = certID
+	return s
+}
+
+func createService(lbID string) *v1.Service {
+	return &v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "test",
+			UID:  "foobar123",
+			Annotations: map[string]string{
+				annDOProtocol:        "http",
+				annoDOLoadBalancerID: lbID,
+			},
+		},
+		Spec: v1.ServiceSpec{
+			Ports: []v1.ServicePort{
+				{
+					Name:     "test",
+					Protocol: "TCP",
+					Port:     int32(443),
+					NodePort: int32(30000),
+				},
+			},
+		},
+	}
+}
+
+func Test_LBaaSCertificateScenarios(t *testing.T) {
+	testcases := []struct {
+		name                  string
+		setupFn               func(fakeLBService, kvCertService) *v1.Service
+		expectedServiceCertID string
+		expectedLBCertID      string
+		err                   error
+	}{
+		{
+			name: "default test values, tls not enabled",
+			setupFn: func(lbService fakeLBService, certService kvCertService) *v1.Service {
+				lb := createLB()
+				lbService.store[lb.ID] = lb
+				return createService(lb.ID)
+			},
+		},
+
+		// lets_encrypt test cases
+		{
+			name: "[letsencrypt] LB cert ID and service cert ID match ",
+			setupFn: func(lbService fakeLBService, certService kvCertService) *v1.Service {
+				lb, cert := createHTTPSLB(443, 30000, "test-lb-id", "test-cert-id", certTypeLetsEncrypt)
+				lbService.store[lb.ID] = lb
+				certService.store[cert.ID] = cert
+				return createServiceWithCert(lb.ID, cert.ID)
+			},
+			expectedServiceCertID: "test-cert-id",
+			expectedLBCertID:      "test-cert-id",
+		},
+		{
+			name: "[letsencrypt] LB cert ID and service cert ID differ and both certs exist",
+			setupFn: func(lbService fakeLBService, certService kvCertService) *v1.Service {
+				lb, cert := createHTTPSLB(443, 30000, "test-lb-id", "test-cert-id", certTypeLetsEncrypt)
```

Comment:
nit: would name `test-cert-id` as `lb-cert-id` as "test" has fairly little meaning.

Reply:
Ah yeah, I meant to do this eventually but forgot -- I was originally following established naming conventions I saw in the loadbalancer tests.